File: frontend/src/types/demand.ts
// demand.ts
export interface BlockDemand {
  id: number;
  demand_code: string;
  source_system: 'TMS' | 'SMMS' | 'TDMS';
  department: 'P_WAY' | 'OHE' | 'S_AND_T';
  section_from: string;
  section_to: string;
  start_km: number;
  end_km: number;
  requested_date: string;
  requested_start_minutes: number;
  requested_end_minutes: number;
  required_minutes: number;
  activity_description: string;
  machinery_type?: string;
  machinery_id?: string;
  status: 'PROPOSED' | 'REVIEWED' | 'APPROVED' | 'EXECUTED' | 'CLOSED' | 'REJECTED';
  trust_score: number;
  severity_tier: 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW';
  priority_weight: number;
  power_block_required: boolean;
  disconnection_required: boolean;
  speed_restriction_kmph?: number;
}

File: backend/app/models/demand.py
from sqlalchemy import Column, Integer, String, Float, Boolean, Date, Text
from app.core.database import Base

class DemandModel(Base):
    __tablename__ = "block_demands"

    id = Column(Integer, primary_key=True, index=True)
    demand_code = Column(String(50), unique=True, index=True, nullable=False)
    source_system = Column(String(20), nullable=False)  # TMS, SMMS, TDMS
    department = Column(String(20), nullable=False)     # P_WAY, OHE, S_AND_T
    section_from = Column(String(10), nullable=False)
    section_to = Column(String(10), nullable=False)
    start_km = Column(Float, nullable=False)
    end_km = Column(Float, nullable=False)
    requested_date = Column(String(20), nullable=False)
    requested_start_minutes = Column(Integer, nullable=False)
    requested_end_minutes = Column(Integer, nullable=False)
    required_minutes = Column(Integer, nullable=False)
    activity_description = Column(Text, nullable=False)
    machinery_type = Column(String(50), nullable=False)
    machinery_id = Column(String(50), nullable=True)
    status = Column(String(20), default="PROPOSED", nullable=False)  # PROPOSED, REVIEWED, APPROVED, EXECUTED, CLOSED
    trust_score = Column(Float, default=90.0)
    severity_tier = Column(String(20), default="HIGH")              # CRITICAL, HIGH, MEDIUM, LOW
    priority_weight = Column(Float, default=7.5)
    power_block_required = Column(Boolean, default=False)
    disconnection_required = Column(Boolean, default=False)
